checker-util/src/main/java/org/checkerframework/checker/regex/util/RegexUtil.java

```diff
@@ -362,57 +398,128 @@ private static int getGroupCount(Pattern p) {
     }
 
     /**
-     * Return the non-null groups in the argument.
+     * Returns a list of groups other than 0, that are guaranteed to be non-null given that the
+     * regular expression matches the String.
      *
-     * @param p pattern to be analysed
+     * @param regexp pattern to be analysed
+     * @param n number of capturing groups in the pattern
      * @return the groups that are non-null in the argument
      */
-    private static List<Integer> getNonNullGroups(Pattern p) {
-        String regexp = p.pattern();
-        int n = getGroupCount(p);
+    public static List<Integer> getNonNullGroups(String regexp, int n) {
         List<Integer> nonNullGroups = new ArrayList<>();
-        for (int i = 0; i <= n; i++) {
+        // Initialize the list with all elements; the below code will remove some.
+        for (int i = 1; i <= n; i++) {
             nonNullGroups.add(i);
         }
+
         ArrayDeque<Integer> openingIndices = new ArrayDeque<>();
+        ArrayDeque<Boolean> openingWasGroup = new ArrayDeque<>();
```

Comment:
Document these variables.  (I see `openingWasGroup` documented below, but not referenced by name, and there should be at least a brief comment here.)